File: Codes/Programmers/키패드누르기.py
def dist(x, y):
    return abs(x[0] - y[0]) + abs(x[1] - y[1])


def solution(numbers, hand):
    answer = ""

    locations = {0: (3, 1)}
    for i in range(1, 10):
        q, r = divmod(i - 1, 3)
        locations[i] = (q, r)

    left, right = (3, 0), (3, 2)

    left_col = {1, 4, 7}
    right_col = {3, 6, 9}

    for num in numbers:
        if num in left_col:
            answer += "L"
            left = locations[num]
        elif num in right_col:
            answer += "R"
            right = locations[num]
        else:
            if dist(left, locations[num]) > dist(right, locations[num]):
                answer += "R"
                right = locations[num]
            elif dist(left, locations[num]) < dist(right, locations[num]):
                answer += "L"
                left = locations[num]
            elif hand == "left":
                answer += "L"
                left = locations[num]
            else:
                answer += "R"
                right = locations[num]

    return answer

File: Codes/Programmers/키패드누르기.js
function solution(numbers, hand) {
  const getDist = (posA, posB) => {
    return Math.abs(posA[0] - posB[0]) + Math.abs(posA[1] - posB[1]);
  };

  const handPos = {
    left: [3, 0],
    right: [3, 2],
  };

  const result = numbers.map((num) => {
    if (num % 3 === 1) {
      handPos.left = [parseInt(num / 3), 0];
      return "L";
    } else if (num % 3 === 2 || num === 0) {
      const pos = num !== 0 ? [parseInt(num / 3), 1] : [3, 1];
      const D = getDist(pos, handPos.left) - getDist(pos, handPos.right);
      if (D > 0) {
        handPos.right = pos;
        return "R";
      } else if (D < 0) {
        handPos.left = pos;
        return "L";
      } else if (hand === "left") {
        handPos.left = pos;
        return "L";
      } else {
        handPos.right = pos;
        return "R";
      }
    } else {
      handPos.right = [parseInt(num / 3) - 1, 2];
      return "R";
    }
  });

  return result.join("");
}
